docs/advanced/cast/custom.rst

```diff
@@ -31,63 +31,86 @@ The following Python snippet demonstrates the intended usage from the Python sid
+         */
+        PYBIND11_TYPE_CASTER(inty, const_name("inty"));
+        /**
+         * Conversion part 1 (Python->C++): convert a PyObject into a inty
+         * instance or return false upon failure. The second argument
+         * indicates whether implicit conversions should be applied.
+         */
+        bool load(handle src, bool) {
+            /* Extract PyObject from handle */
+            PyObject *source = src.ptr();
+            /* Try converting into a Python integer value */
+            PyObject *tmp = PyNumber_Long(source);
+            if (!tmp)
+                return false;
+            /* Now try to convert into a C++ int */
+            value.long_value = PyLong_AsLong(tmp);
+            Py_DECREF(tmp);
+            /* Ensure return code was OK (to avoid out-of-range errors etc) */
+            return !(value.long_value == -1 && !PyErr_Occurred());
+        }
+        /**
+         * Conversion part 2 (C++ -> Python): convert an inty instance into
+         * a Python object. The second and third arguments are used to
+         * indicate the return value policy and parent object (for
+         * ``return_value_policy::reference_internal``) and are generally
+         * ignored by implicit casters.
+         */
+        static handle cast(inty src, return_value_policy /* policy */, handle /* parent */) {
+            return PyLong_FromLong(src.long_value);
+        }
+    };
 
 .. note::
 
-    A ``type_caster<T>`` defined with ``PYBIND11_TYPE_CASTER(T, ...)`` requires
+    A caster class defined with ``PYBIND11_TYPE_CASTER(T, ...)`` requires
     that ``T`` is default-constructible (``value`` is first default constructed
     and then ``load()`` assigns to it).
 
-.. warning::
+The caster defined above must be registered with pybind11.
+There are two ways to do it:
 
-    When using custom type casters, it's important to declare them consistently
-    in every compilation unit of the Python extension module. Otherwise,
-    undefined behavior can ensue.
+* As an instantiation of the ``pybind11::detail::type_caster<T>`` template.
+  Although this is an implementation detail, adding an instantiation of this
+  type is explicitly allowed:
+
+  .. code-block:: cpp
+
+      namespace pybind11 { namespace detail {
+          template <> struct type_caster<inty> : inty_type_caster {};
+      }} // namespace pybind11::detail
+
+  .. warning::
+
+      When using this method, it's important to declare them consistently
+      in every compilation unit of the Python extension module. Otherwise,
+      undefined behavior can ensue.
+
+* When you own the namespace where the type is defined, the preferred method
+  is to *declare* a function named ``pybind11_select_caster``, its only purpose
+  is to associate the C++ type with its caster class:
+
+  .. code-block:: cpp
+
+      struct inty_type_caster;
+
+      struct inty {
+        ...
+        friend inty_type_caster pybind11_select_caster(inty*);
```

Comment:
This looks like the feature is intrusive on the wrapped C++ code, it is not immediately obvious that that's a choice. I think the previous version of the doc was actually better, because it makes it clear that `pybind11_select_caster` is usually only in the pybind11 bindings code (not user code). Maybe adding a note could nice:

Note that `pybind11_select_caster` can also be implemented as a `friend` function if that is preferred.

Reply:
I think that it's preferred to use it as a friend function. But it is permitted to be used as a non-friend function.

Reply:
> I think that it's preferred to use it as a friend function.

My main worry is that people walk away with the wrong impression that this feature is intrusive on their C++ code, and therefore don't use `pybind11_select_caster` at all. With the doc as is, I'd bet that's fairly likely to happen.

Reply:
I agree. Maybe a small comment stating it is not required if you can use only public variables?